clang_build/project.py

```diff
@@ -154,104 +176,145 @@ def __init__(self, directory, environment, **kwargs):
-        Returns
-        -------
-        list
-            List of TargetDescription objects
-        """
-        n_targets = len([None for key, val in self.config.items() if isinstance(val, dict)])
-        only_target = n_targets == 1 and len(self.config.get("subprojects", [])) == 0
-        return [
-            _TargetDescription(
-                key, val, self._identifier_from_name(key), self, self.environment, only_target=only_target
-            )
-            for key, val in self.config.items()
-            if isinstance(val, dict)
-        ]
+        return cls.from_config(config, directory, environment, **kwargs)
 
-    def _check_for_circular_dependencies(self):
-        """Check if targets are defined in a circular dependency.
+    @classmethod
+    def from_config(cls, config: dict, directory, environment, **kwargs):
+        """Generate a project from a config dictionary.
 
-        Raises an exception if circular dependencies were found pointing
-        out where the circular dependencies were found.
+        optional kwarg `targets`: Any of the `targets` may be a `Target` or `TargetDescription`
         """
-        circles = list(_nx.simple_cycles(self._project_tree))
+        name = str(config.get("name", ""))
+        project = cls(name, config, directory, environment, **kwargs)
 
-        circles = [_Circle(circle + [circle[0]]) for circle in circles]
-        if circles:
-            error_message = "Found the following circular dependencies:\n" + _textwrap.indent(
-                    "\n".join("- " + str(circle) for circle in circles), prefix=" " * 3)
-            self._logger.exception(error_message)
-            raise RuntimeError(self.log_message(error_message))
+        return project
+
+    def add_targets(self, target_list: list):
+        """Add a list of targets to this project.
 
-    def _fill_dependency_tree(self):
-        """Integrate this project into the global project tree.
+        This method does the following:
+        - check the validity of targets in the given list
+        - add to the project's target_list
+        - amend the dependency tree:
+          - Add nodes and edges for targets in self
+          - Add edges for dependencies in targets defined in project
+        - optionally create a new dotfile
+        - check for circular dependencies in the project
+
+        This method integrates this project into the global project tree.
 
         This helper function is part of the initialisation of a project.
         It adds this project, all targets and their dependencies to the
         global project tree. If there are illegal dependencies, this function
         will raise an exception.
         """
-        # Add self
-        self._project_tree.add_node(self, data=self)
+        for target in target_list:
+            if isinstance(target, _Target):
+                raise RuntimeError(f"clang_build.project.Project.add_targets: cannot add non-`Target` instance of type {type(target)}")
```

Comment:
Is the "non-" a mistake by any chance?

Reply:
Good that you noticed - actually, the `target` should be (derived from) `TargetDescription` or `Target`, so I will change this check.